### ISSUE TITLE
"column(s) values are all na" error 

### ISSUE THREAD

**Issue by e1937671** [twopirllc/pandas-ta#196]
**Which version are you running? The lastest version is on Github. Pip is for major releases.**
```python
import pandas_ta as ta
print(ta.version)
0.2.28b0
```

**Upgrade.**
```sh
$ pip install -U git+https://github.com/twopirllc/pandas-ta
```

**Describe the bug**
I'm trying to run ta.strategy(verbose = True) in Jupyter Notebook in a local Ubuntu machine which is controlled by SSH connection, and I get "column(s) values are all na: SUPERTl_7_3.0" messages and an error. When I call indicators with CustomStrategy, it works well and I can get all indicators with no error. 

When I try ta.strategy() in Google Cloud Platform, it works very well. Therefore; I think the problem is related to my local machine. Maybe some problems with multiprocessing or versions etc.; but couldn't figure out the exact problem so far. Do you have any suggestions? I uploaded my dataset and code so that you can look at them. I'm using the latest version of pandas_ta. 
[inputs.zip](https://github.com/twopirllc/pandas-ta/files/5860064/inputs.zip)

**To Reproduce**
```python
import os 
import pandas as pd
import numpy as np
import pandas_ta as ta

inputs = pd.read_csv('inputs.csv')

inputs.ta.strategy()`
```

**Expected behavior**
I'm expecting ta.Strategy() to operate with no error and give the proper indicators. Indeed, I didn't get any error when I'm working in Google Cloud Platform. 

**Screenshots**
The first 2 screenshots (dark theme) are from the error case in the local machine, and the 3rd screenshoot (light theme) is from the Google Cloud Platform, in which there is no error. 
![image](https://user-images.githubusercontent.com/26437597/105555046-a6dd0400-5d19-11eb-8163-4f0d58c957dc.png)
![image](https://user-images.githubusercontent.com/26437597/105555086-b6f4e380-5d19-11eb-9309-04f4816de248.png)
![image](https://user-images.githubusercontent.com/26437597/105576357-408ccb80-5d83-11eb-91b5-4926c3eb4b3b.png)

Thanks for using Pandas TA!

**Comment by twopirllc** [comment]
Hello @e1937671,

Thanks for making this a separate Issue and providing me with more detail so I can help you more accurately.

Weird, I do not know why it works on the Google Cloud Platform but not on a local machine; I assume you are using the exact same code.

I recommend upgrading to the latest version _0.2.40b_, which I just uploaded to Github yesterday.

```sh
$ pip install -U git+https://github.com/twopirllc/pandas-ta
```

There was a fix to **vwap** so it can be anchored to Days, Weeks, etc. but you have to set the DataFrame index to be a DatetimeIndex as shown below.


```python
import pandas as pd
import pandas_ta as ta

inputs = pd.read_csv("inputs.csv")
inputs.set_index(pd.DatetimeIndex(inputs["datetime"]), inplace=True)
inputs = inputs[inputs.columns[-5:]]
inputs.ta.strategy(verbose=True)
```

Hope this helps! Let me know if this works for you.

Regards,
KJ

**Comment by twopirllc** [comment]
Hello @e1937671,

Just following up since I have not heard seen a response on this Issue. Have any new information or progress?

Thanks,
KJ

**Comment by e1937671** [comment]
Hello @twopirllc 

Firstly, thank you for your answer. Actually, we can only solve the problem by excluding Supertrend and we can get an output with no error. But when we do this, we see other problems. For example, when we run Strategy and count the number of NaN's, log_return gives too many NaN's. 
![image](https://user-images.githubusercontent.com/26437597/105898350-5ec82500-602a-11eb-83c4-56e5a7d94153.png)
However, when we call log_return alone, there is no problem. 
![image](https://user-images.githubusercontent.com/26437597/105899135-558b8800-602b-11eb-82bc-7d63546f5af0.png)


In another example, the WMA column is totally different in ta.Strategy and ta.wma.. 

![image](https://user-images.githubusercontent.com/26437597/105899917-61c41500-602c-11eb-85f0-2f51415908e6.png)

And when we run ta.wma() the output is the correct one.
![image](https://user-images.githubusercontent.com/26437597/105900091-a2239300-602c-11eb-901a-8db600fda5bb.png)


So I believe there is something related to Strategy() and our setup. Maybe some compatibility issue, we don't know yet. But the problem is related to our Ubuntu setup because we didn't experience this issue in Google Cloud Platform. 

In addition, out machine doesn't run the multiprocessing even though "Multiprocessing: 128 of 128 cores" is written and the process takes around too much time compared to our expectations.( We opened another thread for this issue therefore I won't go in detail.) (https://github.com/twopirllc/pandas-ta/issues/198) In this step, I decided try to run multiprocessing manually while calling all the indicators without using Strategy. Then we will see if we can improve something. (Definitely we will get correct indicator outputs but don't know if we improve the speed)

My theory is, there is something missing in our setup and it affects the working of Strategy(), and causes both multiprocessing and Supertrend issue. But for now, we couldn't get any clue. We are working on :) If you can get any idea, you can share with us :)

Thanks for the help,
M. Burakhan Yildiz

**Comment by M6stafa** [comment]
Hi, i had this problem too. in my case, after i debug the code, i figure out that if you have a window that prices only go higher or lower (in small windows this can happen usually), only SUPERTl or SUPERTs has value and the other one is all NaN; and because of this [line](https://github.com/twopirllc/pandas-ta/blob/master/pandas_ta/core.py#L363), for every other indicator that will add after supertrend, i got this error.

For me, this error happend for "supertrend", "psar", "qqe" and "hilo". All of them has the same concept of separate columns for long and short.

My solution was use fillna for these four indicator, like: `dict(kind='supertrend', fillna=0)` in stratagy ta list.

Finally i don't know why you added that line of check NaN? can you explain the situation you were thinking of?

**Comment by e1937671** [comment]
Hi @M6stafa . Thanks for your reply. I wanted to check the number of NaN's in each indicator, so that whether I can use it as a proper feature. The first 30-40-50 rows with NaN's are not a problem. They are due to windowing. But if there are indicators with a huge number of NaN's, (e.g. 1000-2000) I should exclude them because I'm using dropna after getting indicators, and these columns reduce the data with a huge amount. And yes, I had similar problems with "pasr, qqe and hilo" in another cases and I had also excluded them too. 

Your suggestion sounds good and I will try it and will back here if there is an improvement. But the problem is; I didn't have any problem in Jupyter Notebook in Google Cloud Platform. Same dataset, same code, but different results in different environments.

**Comment by twopirllc** [comment]
Hello @e1937671 and @M6stafa,

### NaN gaps of some indicators

> Hi, i had this problem too. in my case, after i debug the code, i figure out that if you have a window that prices only go higher or lower (in small windows this can happen usually), only SUPERTl or SUPERTs has value and the other one is all NaN; and because of this [line](https://github.com/twopirllc/pandas-ta/blob/master/pandas_ta/core.py#L366), for every other indicator that will add after supertrend, i got this error.

> Finally i don't know why you added that line of check NaN? can you explain the situation you were thinking of?

The reason for the NaNs for things like "supertrend", "psar", "qqe" and "hilo" is for plotting purposes; it messes up the charting. Also the [line](https://github.com/twopirllc/pandas-ta/blob/master/pandas_ta/core.py#L366) is there to not append columns that have all NaNs as referenced and fixed in Issue #97.

> My solution was use fillna for these four indicator, like: dict(kind='supertrend', fillna=0) in stratagy ta list.

Right. This is a good approach for pre-processing your TA output prior to running AI/ML. This is a TA library and not a AI/ML library so it is on the user to handle post-processing. As you pointed out, it takes a line (or a few) to clean up prior to AI/ML processing.

### NaNs of this Issue

> The first 30-40-50 rows with NaN's are not a problem. They are due to windowing. But if there are indicators with a huge number of NaN's, (e.g. 1000-2000) I should exclude them because I'm using dropna after getting indicators, and these columns reduce the data with a huge amount.

I see this also on my local machine with ```log_return``` and others previously mentioned above when running the local strategy test. Just to clarify, the ```log_return``` issue does not happen with Google Cloud Platform, only command line/ terminal?

What I have noticed (and this is difficult to debug due to the nature of multiprocessing) is that somehow the original DataFrame's _ohlcv_ columns are somehow getting mangled before indicators like *log_return()* runs them thus creating all the NaNs. It's possible this may be an OS/architecture Issue. This also started happening on my new Apple M1 MacBook Pro and not my older 2013 MacBook Pro. So it needs a deeper dive.

Hopefully this provides more insight and we all can solve this Issue together.

Thanks,
KJ

**Comment by M6stafa** [comment]
> The reason for the NaNs for things like "supertrend", "psar", "qqe" and "hilo" is for plotting purposes; it messes up the charting. Also the [line](https://github.com/twopirllc/pandas-ta/blob/master/pandas_ta/core.py#L366) is there to not append columns that have all NaNs as referenced and fixed in Issue #97.

But you add the column with all NaNs and not append every other column after it. You check for currently appended columns, not for the new column.
Anyway, this should be an option to filter all NaNs columns or not.

> What I have noticed (and this is difficult to debug due to the nature of multiprocessing) is that somehow the original DataFrame's _ohlcv_ columns are somehow getting mangled before indicators like _log_return()_ runs them thus creating all the NaNs. It's possible this may be an OS/architecture Issue. This also started happening on my new Apple M1 MacBook Pro and not my older 2013 MacBook Pro. So it needs a deeper dive.

Hmm, maybe on the google cloud, supertrend finished after all other indicators.

@e1937671 can you run the code with [this version](https://github.com/M6stafa/pandas-ta/tree/development). Set the `df.ta.cores = 0` so the strategy run without multiprocessing. Test this on both devices and they should have the same result.

**Comment by e1937671** [comment]
Hi @twopirllc . Yes, in Google Cloud Platform, there is not any problem.. No NaNs , no log_return isseus, no improper values for some moving averages... Everything work perfectly. (of course there are NaNs in the first few rows due to windowing)

My initial theory was some compatibility, architecture or version issue too.. So if you experienced the same problem, I guess we got to the point I thought :)

**Comment by twopirllc** [comment]
Hi @e1937671,

> Yes, in Google Cloud Platform, there is not any problem..

Well that's good. Though I understand it is currently an undesired solution. But we certainly need to solve this so we can avoid notebooks.

Out of curiosity, how long does it take to run one ticker will ```all``` the indicators in a Jupyter Notebook? It also runs fine on my new Mac in a Jupyter Notebook with 5344 initial rows/observations.

<img width="830" alt="Screen Shot 2021-01-27 at 6 16 49 PM" src="https://user-images.githubusercontent.com/18198392/106080556-25e08c80-60cc-11eb-85e6-91ff05d87d23.png">

> My initial theory was some compatibility, architecture or version issue too.. So if you experienced the same problem, I guess we got to the point I thought :)

Yeah. There is no version Issue with Pandas TA. 

Have you guys tried the [Linux Python Multiprocessing Issue](https://www.wikitechy.com/tutorials/linux/why-does-multiprocessing-use-only-a-single-core-after-i-import-numpy) that I recommended to @myilmazhaubtech in Issue #198?

Thanks,
KJ

**Comment by myilmazhaubtech** [comment]
Yes we tried the parallel code on the website you mentioned it worked perfectly fine on our server. Below is the screen snapshot:

![image](https://user-images.githubusercontent.com/77806252/106176802-c46cfc00-6165-11eb-93b9-b61a54ee9ee9.png)

**Comment by twopirllc** [comment]
Hi @myilmazhaubtech,

But Pandas TA is still not working

KJ

**Comment by myilmazhaubtech** [comment]
Unfortunately not in parallel. Each ticker takes almost half a minute to run using the `pandas-ta` on an AMD Ryzen Threadripper 3990X processor. However similar number of indicators take about 2 seconds with the old ta-lib library. And then there is the data quality problem. 

I am afraid we will have to wait until these bugs are cleaned.

**Comment by twopirllc** [comment]
@myilmazhaubtech,

Did you try any of the **three** solutions in your code? Not just running some sample code?

For instance: 
<img width="689" alt="Screen Shot 2021-01-28 at 10 12 37 AM" src="https://user-images.githubusercontent.com/18198392/106180700-81089280-6151-11eb-8086-f8e278b19e77.png">

> And then there is the data quality problem.

What data quality?

> I am afraid we will have to wait until these bugs are cleaned.

Whose bugs?

**Comment by twopirllc** [comment]
Hello @e1937671 and @myilmazhaubtech,

> My suggestion would be to remove the multiprocessing functionality altogether and leave that part to the user. I really don't see any reason for this rich library to malfunction other than the suspect multiprocessing feature. It is really a breeze to add so many columns with a single function call here.

Thanks for the feedback!

Kind Regards,
KJ
